trieste/pareto.py

```diff
@@ -0,0 +1,132 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" This module contains functions and classes for Pareto based multi-objective optimization. """
+
+from typing import Tuple
+
+import tensorflow as tf
+from typing_extensions import Final
+
+from .type import TensorType
+
+
+def non_dominated(observations: TensorType) -> Tuple[TensorType, TensorType]:
+    """
+    Computes the non-dominated set for a set of data points.
+    if there are duplicate point(s) in the non-dominated set, this function will return
+    as it is without removing the duplicate.
+
+    :param observations: set of points with shape [N,D]
+    :return: tf.Tensor of the non-dominated set [P,D] and the degree of dominance [N],
+        P is the number of points in pareto front
+        dominances gives the number of dominating points for each data point
+
+
+    """
+    extended = tf.tile(observations[None], [len(observations), 1, 1])
+    swapped_ext = tf.transpose(extended, [1, 0, 2])
+    dominance = tf.math.count_nonzero(
+        tf.logical_and(
+            tf.reduce_all(extended <= swapped_ext, axis=2),
+            tf.reduce_any(extended < swapped_ext, axis=2),
+        ),
+        axis=1,
+    )
+
+    return tf.boolean_mask(observations, dominance == 0), dominance
+
+
+class BoundedVolumes:
+    """
+    A :class:`BoundedVolumes store the index of the Pareto front to form lower and upper
+    bounds of the pseudo cells decomposition.
+    """
+
+    def __init__(self, lower_idx: tf.Tensor, upper_idx: tf.Tensor):
+        """
+        Construct bounded volumes.
+
+        :param lower_idx: the lowerbounds index of the volumes
+        :param upper_idx: the upperbounds index of the volumes
+        """
+
+        assert tf.reduce_all(lower_idx.shape == upper_idx.shape)
+        self.lower_idx = lower_idx
+        self.upper_idx = upper_idx
+
+    def append(self, lower_idx: tf.Tensor, upper_idx: tf.Tensor):
+        """
+        Add new bounded volumes.
+
+        :param lower_idx: the lowerbounds of the volumes
+        :param upper_idx: the upperbounds of the volumes
+        """
+        self.lower_idx = tf.concat([self.lower_idx, lower_idx], axis=0)
+        self.upper_idx = tf.concat([self.upper_idx, upper_idx], axis=0)
+
+    def clear(self):
+        # Clears all stored bounded volumes
+        dtype = self.lower_idx.dtype
+        outdim = self.lower_idx.shape[1]
+        self.lower_idx = tf.zeros([0, outdim], dtype=dtype)
+        self.upper_idx = tf.zeros([0, outdim], dtype=dtype)
+
+    def size(self) -> tf.Tensor:
+        """
+        :return: volume of each bounded volume
+        """
+        return tf.math.reduce_prod(self.upper_idx - self.lower_idx, axis=0, keepdims=True)
```

Comment:
these three methods aren't used in this PR. Are they needed yet?

Reply:
not yet for this PR, might be used in the divide and conquer method. I'll remove it

Reply:
removed